packages/mobile/src/screens/search-screen-v2/screens/FilterBpmScreen.tsx

```diff
@@ -0,0 +1,338 @@
+import { useCallback, useEffect, useMemo, useState } from 'react'
+
+import { Button, Flex, Text, TextInput, useTheme } from '@audius/harmony-native'
+import { SegmentedControl } from 'app/components/core'
+import { FormScreen } from 'app/screens/form-screen'
+import { ListSelectionItems } from 'app/screens/list-selection-screen/ListSelectionItems'
+
+import { useSearchFilter } from '../searchState'
+
+const MIN_BPM = 1
+const MAX_BPM = 999
+
+const messages = {
+  title: 'Bpm',
+  bpm: 'BPM',
+  range: 'Range',
+  target: 'Target',
+  minBpm: 'Min',
+  maxBpm: 'Max',
+  tooLowError: `BPM less than ${MIN_BPM}`,
+  tooHighError: `BPM greater than ${MAX_BPM}`,
+  invalidMinMaxError: 'Min greater than max'
+}
+
+type ViewProps = {
+  value: string | undefined
+  setValue: (value: string) => void
+}
+
+type BpmTargetType = 'exact' | 'range5' | 'range10'
+const targetOptions: { label: string; value: BpmTargetType }[] = [
+  {
+    label: 'Exact',
+    value: 'exact'
+  },
+  {
+    label: '± 5',
+    value: 'range5'
+  },
+  {
+    label: '± 10',
+    value: 'range10'
+  }
+]
+
+const rangeOptions = [
+  {
+    label: 'Very Slow (<60 BPM)',
+    value: `${MIN_BPM}-60`
+  },
+  {
+    label: 'Slow (60-90 BPM)',
+    value: '60-90'
+  },
+  {
+    label: 'Medium (90-110 BPM)',
+    value: '90-110'
+  },
+  {
+    label: 'Upbeat (110-140 BPM)',
+    value: '110-140'
+  },
+  {
+    label: 'Fast (140-160 BPM)',
+    value: '140-160'
+  },
+  {
+    label: 'Very Fast (160+ BPM)',
+    value: `160-${MAX_BPM}`
+  },
+  {
+    label: 'Custom',
+    value: 'custom'
+  }
+]
+
+const BpmRangeView = ({ value, setValue }: ViewProps) => {
+  const [bpmRange, setBpmRange] = useState<string>()
+  const [minBpm, setMinBpm] = useState('')
+  const [maxBpm, setMaxBpm] = useState('')
+  const [minError, setMinError] = useState<string | null>(null)
+  const [maxError, setMaxError] = useState<string | null>(null)
+  // NOTE: Memo to avoid the constantly changing function instance from triggering the effect
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const onChange = useMemo(() => setValue, [])
+
+  useEffect(() => {
+    if (bpmRange === 'custom') {
+      setValue('')
+    } else if (bpmRange) {
+      onChange(bpmRange)
+    }
+  }, [bpmRange, onChange, setValue])
+
+  useEffect(() => {
+    if (!value && (bpmRange || minBpm || maxBpm)) {
+      if (bpmRange !== 'custom') setBpmRange('')
+      setMinBpm('')
+      setMaxBpm('')
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value, setBpmRange, setMinBpm, setMaxBpm])
+
+  useEffect(() => {
```

Comment:
hmm this does seem quite complicated... maybe a helper, but also im surprised there is this much validation needed. would it make sense to move all this into "MinBpmTextInput" and "MaxBpmTextInput"? could we use zod + a lil formik action also?

Reply:
Yea this got a little trickier bc i'm effectively trying to sync three different levels of state. I might leave this for this PR bc things are working, but I'll make a note to try to simplify this code a bit